### PR TITLE
Make sure the TLS peer chain is always a true empty stack

### PR DESCRIPTION
**Pull request by levitte**
##### Checklist

- [x] CLA is signed

### CODE REVIEW

**Comment by levitte** [comment]
This should answer the key issue in #2254

**Comment by levitte** [comment]
If this is an appropriate fix (I'm not sure, see my comments in #2254), then I'll also prepare a backport to 1.0.2.

**Comment by triska** [comment]
Thank you for working on this!

Please also think of the manual page!

File: ssl/ssl_sess.c [hunk 1]
```diff
@@ -92,18 +92,24 @@ SSL_SESSION *SSL_SESSION_new(void)
...
 
     if (!CRYPTO_new_ex_data(CRYPTO_EX_INDEX_SSL_SESSION, ss, &ss->ex_data)) {
-        CRYPTO_THREAD_lock_free(ss->lock);
-        OPENSSL_free(ss);
-        return NULL;
+        goto err;
     }
     return ss;
+ err:
+    if (ss) {
```
Comment (dot-asm) [inline ssl/ssl_sess.c]: All `goto err` are performed after ss is dereferenced [multiple times]. Hence if (ss) is redundant and becomes unjustified load on branch prediction logic.

Comment (levitte) [inline ssl/ssl_sess.c]: Good point.  Will fix

Comment (levitte) [inline ssl/ssl_sess.c]: Done

**Comment by levitte** [comment]
Unfortunately, I found out that this change must be post-poned until 1.2.0.  It seems that NULL is an expected return when no certificates were sent by the peer.

**Comment by levitte** [comment]
This was postponed...  now's the time for a review, though.

**Comment by kaduk** [comment]
I feel like there ought to be a CHANGES entry every time we are making API-breaking changes that had to wait for a major version boundary.

**Comment by levitte** [comment]
@kaduk, didn't see your comment before.  I'm not sure I understand how this is an API breaking change, would you mind explaining (if you still remember, this *was* more than half year ago after all)

**Comment by kaduk** [comment]
@levitte I was probably just responding to the comment at https://github.com/openssl/openssl/pull/2256#issuecomment-275705173 about NULL being an unexpected return value.
Since this is a change that we have to wait for a major release to make (since it might break applications), it seems like we should actually tell applications that the change happened, instead of letting them find out organically.

**Comment by levitte** [comment]
Oh, that.  I totally forgot...

**Comment by mattcaswell** [comment]
I tend to think we should not do this. I suspect this will break various things. It might have been better to have designed things this way in the first place - but I kind of think it is too late to change this behaviour now.

**Comment by levitte** [comment]
Yeah ok, I have no problems giving this one up

**Pull request closed without merging**